### ISSUE TITLE
Question: Pseudo Code

### ISSUE THREAD

**Issue by GAFO** [zyantific/zydis#251]
Hello,

is there a way to generate peuso code with zydis like in IDA ?  
I only seem to find the generation of asm code but not peuso code. 

If its not possible:
Any recomendations (libs) which work well with zydis ?

Greetings GAFO

**Comment by flobernd** [comment]
Hi there, what you are looking for, sadly, is out of scope of this library:

1. Decode single instruction (Zydis)
2. Disassemble code blocks with multiple instructions (linear sweep, follow control flow, etc. [like the IDA disassembly view])
3. Pseodo code generation (like the Hex-Rays decompiler)

I'm sadly not aware of any open-source decompiler library, but maybe somebody else has a good answer for you.

May I ask what goal you want to achieve?

**Comment by flobernd** [comment]
Step 2. is not super complicated, if you don't need 100% (that's mostly impossible anyways) correct disassembly. 

Linear sweep (simply disassembling from the beginning of the file or the EP in a loop and always increasing the offset by the length of the previous instruction) is a very easy approach. It has the drawback of getting confused by inline non-code data blocks, but usually after a few more instructions, the offset is aligned again.

Another approach is starting the linear sweep from the EP and all the exported symbols until you encounter a control-flow instruction like JMP, CALL, JNZ, ... Then follow to the destination address and continue the linear sweep until you find another control-flow instruction or RET, etc. This will most likely jump over inline data blocks, but has other drawbacks. For example it's pretty hard to follow indirect branches (`JMP EAX`) and as well there is a possibility to miss certain branches.

Most analysis tools are combining multiple techniques to get most accurate disassembly.

**Comment by GAFO** [comment]
Hm well that might work out if you have access to the whole memory space but im limited to whatever the "target" allows me to read. Guarded regions within sections are a big problem, so I personally would prefer to convert (at least that was the plan) parts of the actual asm code without knowing the surroundings. You see, for any region I need to check first if the memory space is readable and if its not then stop the process where it hits the fence.

So all I can guarantee for the convert mechanism (asm -> pseudo) is the beginning from a function and its end.
Screenshot to that: 
https://prnt.sc/1vfnabv

And that snipped should then end up being pseudo code looking like this: (from IDA)
https://prnt.sc/1vfpi79

So in sum the first goal is to convert small parts and then increase the complexity :)

**Comment by athre0z** [comment]
While step 2 is pretty manageable, step 3 is incredibly difficult. To be completely honest, snowman's output is so terrible that I'd rather not have any pseudocode at all. The only open source tooling that I'm aware of that does a somewhat decent job of decompiling X86 is Ghidra, which will likely not exactly be easy to embed into a different project.

**Comment by GAFO** [comment]
Oh so I probably found the wrong one then, only saw Java files there.
Think this is the correct one now: https://github.com/NationalSecurityAgency/ghidra 
Seems to be a mix of Java and c++ tho - will check in detail then.
